src/langchain_google_spanner/graph_retriever.py

```diff
@@ -0,0 +1,389 @@
+            selector=selector,
+            **kwargs,
+        )
+
+    def add_example(self, question: str, gql: str):
+        if self.selector is None:
+            raise ValueError("`selector` cannot be None")
+        self.selector.add_example(
+            {"input": question, "query": duplicate_braces_in_string(gql)}
+        )
+
+    def _get_relevant_documents(
+        self, question: str, *, run_manager: CallbackManagerForRetrieverRun
+    ) -> List[Document]:
+        """Translate the natural language query to GQL using examples looked up
+        by a semantic similarity model, execute it, and return the results as
+        Documents.
+        """
+
+        if self.llm is None:
+            raise ValueError("`llm` cannot be None")
+        if self.selector is None:
+            raise ValueError("`selector` cannot be None")
+
+        # Define the prompt template
+        prompt = FewShotPromptTemplate(
+            example_selector=self.selector,
+            example_prompt=PromptTemplate.from_template(
+                "Question: {input}\nGQL Query: {query}"
+            ),
+            prefix="""
+            Create an ISO GQL query for the question using the schema.""",
+            suffix=DEFAULT_GQL_TEMPLATE_PART1,
+            input_variables=["question", "schema"],
+        )
+
+        # Initialize the chain
+        gql_chain = prompt | self.llm | StrOutputParser()
+        # 1. Generate gql query from natural language query using LLM
+        gql_query = extract_gql(
+            gql_chain.invoke(
+                {
+                    "question": question,
+                    "schema": self.graph_store.get_schema,
+                }
+            )
+        )
+        print(gql_query)
+
+        # 2. Execute the gql query against spanner graph
+        try:
+            responses = self.graph_store.query(gql_query)[: self.k]
+        except Exception as e:
+            raise ValueError(str(e))
+
+        # 3. Transform the results into a list of Documents
+        documents = []
+        for response in responses:
+            documents.append(convert_to_doc(response))
+        return documents
+
+
+class SpannerGraphNodeVectorRetriever(BaseRetriever):
+    """Retriever that does a vector search on nodes in a SpannerGraphStore.
+    If a graph expansion query is provided, it will be executed after the
+    initial vector search to expand the returned context.
+    """
+
+    graph_store: SpannerGraphStore = Field(exclude=True)
+    embedding_service: Optional[Embeddings] = None
+    label_expr: str = "%"
+    """A label expression for the nodes to search"""
+    return_properties_list: List[str] = []
+    """The list of properties to return"""
+    embeddings_column: str = "embedding"
+    """The name of the column that stores embedding"""
+    query_parameters: QueryParameters = QueryParameters()
+    top_k: int = 3
+    """Number of vector similarity matches to return"""
+    graph_expansion_query: str = ""
```

Comment:
This would be a linear query statement (or statements chained by NEXT) https://cloud.google.com/spanner/docs/graph/queries-overview#composing-larger-queries-next-keyword
This is optional, to expand the context by non-generic queries if requrired.
I'm removing the option for this PR. 
We'll add it back if required later.